### ISSUE TITLE
How should I run the program after building it?

### ISSUE THREAD

**Issue by Recope** [sandialabs/miniGraphics#1]

**Comment by kmorel** [comment]
When you build the code, you will see several programs in the `bin` directory. Each one of these represents a version of a parallel graphics algorithm.

These are all MPI programs. You can simply run them from the command line, but it will only run on one process. It will work, but it will not be parallel (which is somewhat missing the point). To run them in parallel, you typically use `mpirun` or `mpiexec`, whichever comes with the MPI distribution that you used. If you are attempting to run this at an HPC center, different HPC clusters may have different ways of scheduling and launching a parallel job. You'll need to talk to your HPC administrators on how to do this.

Each of these programs accept several command line arguments. Use the `--help` command line argument (e.g., `BinarySwapBase --help` to get usage information.

The point of these programs is to run performance measures rather than do anything otherwise useful. Timing and other information will be written to a file named `timing.yaml` (unless you change the output file with the `--yaml-output` argument).

**Comment by Recope** [comment]
Yes, I noticed what you said. So when I run these programs in parallel using **MPI** there's no way to see the render and interact with the geometry, right? Because right now when I run the program using **MPI**, I can only see **.yaml** production.

**Comment by kmorel** [comment]
That is mostly correct. There is no way to view the rendering result in a window and interact with it. `miniGraphics` is trying to be as simple as possible, and interacting with a windowing system adds a lot of complexity, especially when trying to compile on an HPC system.

However, all the programs have an `--enable-write-image` option that will write out the images they generate to a file. The files are written in [ppm](http://netpbm.sourceforge.net/doc/ppm.html) format, which is a somewhat obscure but easy image format. The files can be shown or converted with [ImageMagick](https://imagemagick.org/index.php) if you don't have another program to read them.

**Comment by Recope** [comment]
> That is mostly correct. There is no way to view the rendering result in a window and interact with it. `miniGraphics` is trying to be as simple as possible, and interacting with a windowing system adds a lot of complexity, especially when trying to compile on an HPC system.
> 
> However, all the programs have an `--enable-write-image` option that will write out the images they generate to a file. The files are written in [ppm](http://netpbm.sourceforge.net/doc/ppm.html) format, which is a somewhat obscure but easy image format. The files can be shown or converted with [ImageMagick](https://imagemagick.org/index.php) if you don't have another program to read them.

I would like to ask one more question, and that is how do I switch the render mode to use OpenGL?

**Comment by kmorel** [comment]
You use the `--paint-opengl` command line option. This option is only available if you turn on the `MINIGRAPHICS_ENABLE_OPENGL` configure option in CMake to `ON`.

**Comment by Recope** [comment]
> You use the `--paint-opengl` command line option. This option is only available if you turn on the `MINIGRAPHICS_ENABLE_OPENGL` configure option in CMake to `ON`.

Yes, When I do this, the program displays an error:
```
. ./../ib/libminiGraphicsCommon . a(MainLoop.cpp.o): In function‘ createPainter(RunOptions const&,
YamlWriter&)' :
MainLoop.cpp:( . text+0x737): undefined reference to、PainteropenGL: :PainteropenGL() '
collect2: error: ld returned 1 exit status
```
And I noticed:
![image](https://user-images.githubusercontent.com/41226982/189472420-35c6cadc-7c1d-43c7-abe0-237602d7b170.png)
Is there something wrong with the process I configured?

**Comment by Recope** [comment]
sorry, I checked the configuration error, and now the OpengL program can be compiled normally.
However, when the program is run, it will still report an error and output a composite of errors.
![image](https://user-images.githubusercontent.com/41226982/189519904-1d42d727-c676-4e94-8f03-38e48216eab1.png)

**Comment by kmorel** [comment]
It's been years since I touched this code, but I compiled the code with OpenGL, and I can replicate the problem. It looks like there is a problem with certain types of OpenGL buffers. The rendering part looks wrong. The main point of this code was to measure the performance of the parallel image reduction part of the process, so the painting part is just there to drive the images that get reduced.

On my system if I add both the `--color-float` and `--depth-none`, the process completes correctly. It looks like floating-point color buffers are working OK, but integer color or floating-point depth buffers are not working.

<img width="602" alt="Screen Shot 2022-09-13 at 5 00 56 PM" src="https://user-images.githubusercontent.com/147933/190008290-3f15a5a8-e2c1-429b-8c49-f34579797f12.png">

If you can suggest a fix for the other types of buffers, I'd be happy to add them.

**Comment by Recope** [comment]
> It's been years since I touched this code, but I compiled the code with OpenGL, and I can replicate the problem. It looks like there is a problem with certain types of OpenGL buffers. The rendering part looks wrong. The main point of this code was to measure the performance of the parallel image reduction part of the process, so the painting part is just there to drive the images that get reduced.
> 
> On my system if I add both the `--color-float` and `--depth-none`, the process completes correctly. It looks like floating-point color buffers are working OK, but integer color or floating-point depth buffers are not working.
> 
> <img alt="Screen Shot 2022-09-13 at 5 00 56 PM" width="602" src="https://user-images.githubusercontent.com/147933/190008290-3f15a5a8-e2c1-429b-8c49-f34579797f12.png">
> 
> If you can suggest a fix for the other types of buffers, I'd be happy to add them.

After commenting out the code in PainterOpenGL. cpp, I got the correct rendering:
`glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA)`
